### PR TITLE
Add xDS test client and server

### PR DESCRIPTION
**Pull request by ericgribkoff**
Draft of the test specification is at https://github.com/grpc/grpc/pull/21856, the automated framework for running the tests (language-agnostic) is in review at https://github.com/grpc/grpc/pull/21869, and the corresponding client and server implementations in Java are [here](https://github.com/grpc/grpc-java/blob/master/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestClient.java) and [here](https://github.com/grpc/grpc-java/blob/master/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java).

### CODE REVIEW

File: test/cpp/interop/xds_interop_client.cc [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+ private:
+  std::unique_ptr<TestService::Stub> stub_;
+};
+
+class LoadBalancerStatsServiceImpl : public LoadBalancerStatsService::Service {
+ public:
+  Status GetClientStats(ServerContext* context,
+                        const LoadBalancerStatsRequest* request,
+                        LoadBalancerStatsResponse* response) {
+    int start_id;
```
Comment (donnadionne) [inline test/cpp/interop/xds_interop_client.cc]: more of a question, how is this server context triggered?  i don't see where LoadBalancerStatsRequest is built and GetClientStats is called

I understand the watcher will wait on the conditional var and then respond.  i am a bit confused how we even get into this server context

File: test/cpp/interop/xds_interop_server.cc [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+using grpc::ServerReader;
+using grpc::ServerReaderWriter;
+using grpc::ServerWriter;
+using grpc::Status;
+using grpc::testing::SimpleRequest;
+using grpc::testing::SimpleResponse;
+using grpc::testing::TestService;
+
+class TestServiceImpl : public TestService::Service {
+  std::string hostname;
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply test/cpp/interop/xds_interop_server.cc]: Done

File: test/cpp/interop/xds_interop_server.cc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  Status UnaryCall(ServerContext* context, const SimpleRequest* request,
+                   SimpleResponse* response) {
+    response->set_server_id(FLAGS_server_id);
+    response->set_hostname(hostname);
+    return Status::OK;
+  }
+};
+
+void RunServer(const int port, std::string hostname) {
+  GPR_ASSERT(port != 0);
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply test/cpp/interop/xds_interop_server.cc]: Done

File: test/cpp/interop/xds_interop_server.cc [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+int main(int argc, char** argv) {
+  grpc::testing::TestEnvironment env(argc, argv);
+  grpc::testing::InitTest(&argc, &argv, true);
+
+  char* hostname = grpc_gethostname();
+  if (hostname == nullptr) {
+    std::cout << "Failed to get hostname, aborting test" << std::endl;
+    return 1;
+  }
+
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply test/cpp/interop/xds_interop_server.cc]: Done

File: test/cpp/interop/xds_interop_client.cc (hunk 1 above)
Comment (ericgribkoff) [reply test/cpp/interop/xds_interop_client.cc]: This is the mechanism for the client to expose LB-related stats during the test. It's a gRPC server exposed by the client process and queried by an external test driver: This is implemented in https://github.com/grpc/grpc/pull/21869 (`run_xds_tests.py`) in OSS and a similar Python-based test driver in our google3 grpclb tests. The driver spins up the server(s) and a client process, manipulates the LB (traffic director) configuration as appropriate, and sends a gRPC request to the `GetClientStats` method to verify client-side LB decisions.

**Pull request merged**